bin/faucet/src/faucet/mod.rs

```diff
@@ -0,0 +1,422 @@
+                })
+            },
+            // TODO: Look into which other errors should be recoverable.
+            //       e.g. Connection error being lost to RPC client is probably not fatal.
+            others => Err(others).context("failed to handle error"),
+        }
+    }
+
+    /// Attempts to rollback back local state to match that indicated by the node.
+    ///
+    /// This relies on parsing the stringified error
+    /// `VerifyTxError::IncorrectInitialAccountCommitment`.
+    ///
+    /// Returns an error if the rollback was unsuccesful. This should be treated as fatal.
+    fn handle_desync(&mut self, err: &str) -> anyhow::Result<()> {
+        let onchain_state = parse_desync_error(err).context("failed to parse desync message")?;
+
+        // Find the matching local state, unless we've dropped it already.
+        let rollback = self
+            .prior_state
+            .iter()
+            .position(|state| state.commitment() == onchain_state)
+            .context("no matching local state to rollback to")?;
+
+        // Rollback the local state.
+        // SAFETY: The index must exist since the element was just found.
+        self.data_store.update_faucet_state(self.prior_state[rollback].clone());
+        self.prior_state.drain(rollback..);
+
+        tracing::warn!(rollback.count = rollback, "desync detected and handled");
+
+        Ok(())
+    }
+
+    /// Fully handles a single request _without_ changing local state.
+    ///
+    /// Caller should update the local state based on the returned result.
+    async fn handle_request(
+        &self,
+        request: MintRequest,
+        rng: impl FeltRng,
+        rpc_client: &mut RpcClient,
+    ) -> MintResult<(AccountDelta, BlockNumber, Note)> {
+        // Generate the payment note and compile it into our transaction arguments.
+        let p2id_note = P2IdNote::build(self.faucet_id(), &request, rng)?;
+        let tx_args = p2id_note.compile()?;
+
+        let tx = self.execute_transaction(tx_args)?;
+        let account_delta = tx.account_delta().clone();
+
+        let tx = Self::prove_transaction(tx)?;
+        let block_height = self.submit_transaction(tx, rpc_client).await?;
+
+        Ok((account_delta, block_height, p2id_note.into_inner()))
+    }
+
+    fn execute_transaction(&self, tx_args: TransactionArgs) -> MintResult<ExecutedTransaction> {
+        // TODO: Is this cheap? Do we need to carry this around with us, or can we just construct
+        //       when needed?
+        TransactionExecutor::new(self.data_store.clone(), Some(self.authenticator.clone()))
+            .execute_transaction(self.id.inner(), BlockNumber::GENESIS, &[], tx_args)
+            .map_err(MintError::Execution)
+    }
+
+    fn prove_transaction(tx: ExecutedTransaction) -> MintResult<ProvenTransaction> {
+        LocalTransactionProver::new(ProvingOptions::default())
+            .prove(tx.into())
+            .map_err(MintError::Proving)
+    }
```

Comment:
I  didn't find time for remote prover support.

We should add a general client for it. Right now the proto definitions are in `base` and I didn't feel motivated to try bring them in here myself.